src/raft.c

```diff
@@ -115,19 +115,185 @@ void entryAttachRaftReq(RedisRaftCtx *rr, raft_entry_t *entry, RaftReq *req)
+ * 3) a shardgroup marked as local (i.e. corresponding to this cluster) that owns the slot as an importing slot and
+ *    has a RaftRedisCommandArray marked as asking
+ */
+
+ShardGroup * GetCommandShardGroup(RedisRaftCtx *rr, RaftRedisCommandArray *cmds)
+{
+    ShardGroup *sg = rr->sharding_info->stable_slots_map[cmds->slot];
+    if (sg != NULL) {
+        return sg;
+    }
+
+    sg = rr->sharding_info->migrating_slots_map[cmds->slot];
+    if (sg && sg->local) {
+        return sg;
+    }
+
+    if (cmds->asking) {
+        ShardGroup *isg = rr->sharding_info->importing_slots_map[cmds->slot];
+        if (isg && isg->local) {
+            return isg;
+        }
+    }
+
+    return sg;
+}
+
+RRStatus validateRaftRedisCommandArray(RedisRaftCtx *rr, RedisModuleCtx *reply_ctx, RaftRedisCommandArray *cmds) {
+    /* Make sure hash slot is mapped and handled locally. */
+    ShardGroup *sg = GetCommandShardGroup(rr, cmds);
+    if (!sg) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "CLUSTERDOWN Hash slot is not served");
+        }
+        return RR_ERROR;
+    }
+
+     if (cmds->slot < 0) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "CLUSTERERROR should not have a negative slot here");
+        }
+        return RR_ERROR;
+    }
+    unsigned int slot = (unsigned int) cmds->slot;
+
+    SlotRangeType slot_type = SLOTRANGE_TYPE_UNDEF;
+    for (size_t i = 0; i < sg->slot_ranges_num; i++) {
+        if (sg->slot_ranges[i].start_slot <= slot && slot <= sg->slot_ranges[i].end_slot) {
+            slot_type = sg->slot_ranges[i].type;
+            break;
+        }
+    }
+
+    if (slot_type == SLOTRANGE_TYPE_UNDEF) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "ERR internal error, couldn't associate a shardgroup slot to this request");
+        }
+        return RR_ERROR;
+    }
+
+    if (!sg->local) {
+        if (reply_ctx) {
+            if (sg->next_redir >= sg->nodes_num) {
+                sg->next_redir = 0;
+            }
+
+            replyRedirect(reply_ctx, cmds->slot, &sg->nodes[0].addr);
+        }
+
+        return RR_ERROR;
+    }
+
+    /* if our keys belong to a local migrating/importing slot, all keys must exist */
+    if (slot_type == SLOTRANGE_TYPE_MIGRATING) {
+        switch (validateKeyExistence(rr, cmds)) {
+            case SomeExist:
+                if (reply_ctx) {
+                    RedisModule_ReplyWithError(reply_ctx, "TRYAGAIN");
+                }
+                return RR_ERROR;
+            case NoneExist:
+                if (reply_ctx) {
+                    replyAsk(rr, reply_ctx, cmds->slot);
+                }
+                return RR_ERROR;
+            case AllExist:
+            case NoKeys:
+                return RR_OK;
+        }
+    } else if (slot_type == SLOTRANGE_TYPE_IMPORTING) {
```

Comment:
else can be removed here, for styling consistency

Reply:
@fadidahanna don't understand?